server/src/main/java/io/crate/planner/operators/JoinPlanBuilder.java

```diff
@@ -337,4 +269,27 @@ private static <V> V removeMatch(Map<Set<RelationName>, V> valuesByNames, Set<Re
         }
         return null;
     }
+
+    /**
+     * Splits the given Symbol tree into a list of correlated sub-queries and a list of remaining symbols.
+     */
+    private record SymbolSplitter(List<Symbol> correlatedSubQueries, List<Symbol> remainder) {
+        private static final SymbolSplitter EMPTY = new SymbolSplitter(List.of(), List.of());
+
+        private static SymbolSplitter of(Symbol from) {
+            if (from == null) {
+                return EMPTY;
+            }
+            var remainder = new ArrayList<Symbol>();
+            var correlatedSubQueries = new ArrayList<Symbol>();
+            for (var symbol : QuerySplitter.split(from).values()) {
+                if (SymbolVisitors.any(s -> s instanceof SelectSymbol x && x.isCorrelated(), symbol)) {
+                    correlatedSubQueries.add(symbol);
+                } else {
+                    remainder.add(symbol);
+                }
+            }
+            return new SymbolSplitter(correlatedSubQueries, remainder);
+        }
+    }
```

Comment:
I would structure it like this:

```
    private record CorrelatedSubqueries(List<Symbol> correlatedSubQueries, List<Symbol> remainder) {}
    
    /**
     * Splits the given Symbol tree into a list of correlated sub-queries and a list of remaining symbols.
     */
    private static CorrelatedSubqueries extractCorrelatedSubqueries(@Nullable Symbol from) {
        if (from == null) {
            return new CorrelatedSubqueries(List.of(), List.of());
        }
        var values = QuerySplitter.split(from).values();
        var remainder = new ArrayList<Symbol>(values.size());
        var correlatedSubQueries = new ArrayList<Symbol>(values.size());
        for (var symbol : values) {
            if (SymbolVisitors.any(s -> s instanceof SelectSymbol x && x.isCorrelated(), symbol)) {
                correlatedSubQueries.add(symbol);
            } else {
                remainder.add(symbol);
            }
        }
        return new CorrelatedSubqueries(correlatedSubQueries, remainder);
    }
```
You can see this record simply as a replacement for Pair.
Minor nitpick: `ArrayList` is initialized with default capacity 10, so going with values.size() will be probably less than 10.

Reply:
Thank you! made the change.